cpp/src/arrow/builder.cc

```diff
@@ -238,6 +238,7 @@ Status PrimitiveBuilder<T>::Init(int64_t capacity) {
   RETURN_NOT_OK(data_->Resize(nbytes));
 
   raw_data_ = reinterpret_cast<value_type*>(data_->mutable_data());
+  memset(raw_data_, 0, nbytes);
```

Comment:
@pcmoritz @alendit this is the memset that fixed the Python determinism unit test. Now I'm trying to figure out why

Reply:
Ok, so some of the initial bytes from the builder are leaking. One possibility I can think of is that we forget to call `Finish` somewhere in one of the builders in `python_to_arrow.cc` and just grab the array, I'm looking into that right now (but haven't found anything so far).

Reply:
@wesm Actually, the types_ builder in SequenceBuilder is never finished, could that be a problem? On the other hand, I think only bytes that have been written by us are extracted from there.

Reply:
Yep, that's the problem.